proposals/NNNN-Promises-in-Policy-Server.md

```diff
@@ -0,0 +1,114 @@
+    const appObjPromises = appObjs.map(async appObj => {
+        const promiseObjects = { // start all promises in parallel
+            categories: asyncSql(sqlApps.getAppCategoriesNames(appObj.id)),
+            displayNames: asyncSql(sql.getAppDisplayNames(appObj.id)),
+            moduleNames: asyncSql(sql.getAppModules(appObj.id)),
+            funcGroupNames: asyncSql(sql.getAppFunctionalGroups(isProduction, appObj)),
+            serviceTypes: asyncSql(sqlApps.getApp.serviceTypes.idFilter(appObj.id)),
+            serviceTypeNames: asyncSql(sqlApps.getApp.serviceTypeNames.idFilter(appObj.id)),
+            serviceTypePermissions: asyncSql(sqlApps.getApp.serviceTypePermissions.idFilter(appObj.id)),
+            hybridPreference: asyncSql(sqlApps.getApp.hybridPreference.idFilter(appObj.id)),
+            appPassthrough: asyncSql(sqlApps.getApp.passthrough.idFilter(appObj.id)),
+            incomingAppPolicy: Promise.resolve(incomingAppObjs[(useLongUuids ? appObj.app_uuid : appObj.app_short_uuid)])
+        };
+
+        for (let prop in promiseObjects) {
+            promiseObjects[prop] = await promiseObjects[prop]; // resolve all promises into each property
+        }
+
+        const allDataPromises = { // start all promises in parallel
+            policyObjs: model.constructAppPolicy(appObj, useLongUuids, promiseObjects),
+            requestedUuids: Promise.resolve(requestedUuids),
+            useLongUuids: Promise.resolve(useLongUuids),
+            defaultFuncGroups: asyncSql(sql.getDefaultFunctionalGroups(isProduction)),
+            preDataConsentFuncGroups: asyncSql(sql.getPreDataConsentFunctionalGroups(isProduction)),
+            deviceFuncGroups: asyncSql(sql.getDeviceFunctionalGroups(isProduction)),
+        };
+        if (requestedUuids.length > 0) {
+            allDataPromises.blacklistedApps = asyncSql(sqlApps.getBlacklistedApps(requestedUuids, useLongUuids));
+        } else {
+            allDataPromises.blacklistedApps = Promise.resolve([]);
+        }
+
+        for (let prop in allDataPromises) {
+            allDataPromises[prop] = await allDataPromises[prop]; // resolve all promises into each property
+        }
+
+        return model.aggregateResults(allDataPromises);
+    });
+
+    return Promise.all(appObjPromises);
+}
+
+
+```
+
+## Potential downsides
+
+Although this is a single conceptual change in the idea of how asynchronous logic should be handled, it will involve a significant editing of the majority of the project's server files due to how ubiquitous the callback-style is. As a result, the project's code may look very different from its current state. All other PRs that edit the server code will be incompatible with the changes made in this PR. Therefore, it is recommended that either no other PR editing the server files is started in conjunction with this one, or a PR can be started but later must get refactored by its author to implement the same changes requested in this proposal.
```

Comment:
```suggestion
Although this is a single conceptual change in the idea of how asynchronous logic should be handled, it will involve significant editing of the majority of the project's server files due to how ubiquitous the callback-style is. As a result, the project's code may look very different from its current state. All other PRs that edit the server code will be incompatible with the changes made in this PR. Therefore, it is recommended that either no other PR editing the server files is started in conjunction with this one, or a PR can be started but later must get refactored by its author to implement the same changes requested in this proposal.
```